aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.amazonaws.services.s3.AmazonS3URI;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.sync.ResponseTransformer;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+
+public class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final S3Client s3;
+  private final AmazonS3URI location;
+
+  private InputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  public S3InputStream(S3Client s3, AmazonS3URI location) {
+    this.s3 = s3;
+    this.location = location;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "already closed");
```

Comment:
Nit: we mostly use sentence case for error messages, and try to include the intent: "Cannot read: already closed"

Reply:
Updated message.